packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,323 @@
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text data-key="name"></nimble-table-column-text>
+    <nimble-table-column-icon data-key="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation. The ordering of the columns in the markup will determine the visual ordering of the columns (top to bottom equals left to right...unless in 'rtl'). Re-ordering of columns will be done, at least at first, through the re-ordering of the columns elements in the DOM.
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    @observable
+    public slottedColumns: TableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where the `TCellData` type is a superset of the type represented by [`TableRecord`](https://github.com/ni/nimble/blob/3e4b8d3dd59431d1671e381aa66052db57bc475c/packages/nimble-components/src/table/types.ts#L24):
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `TableColumn<>`
+
+This abstract class is what a column web component (i.e. a slotted column) must extend.
+
+```TS
+abstract class TableColumn<TCellData extends TableRowData = unknown> {
+    // This method will produce the expected TableCellState that the cellTemplate expects as
+    // its source
+    generateCellState: (cellData: TCellData) => TableCellState<TCellData>;
+
+    // The template to use to render the cell content for the column
```

Comment:
I think @rajsite mused at some point that we'll want a way to return style information too. Is that missing or did you decide it wasn't needed?

Reply:
If there is some relevant style information we want to return we can decide that at a later time. For now, I don't see a pressing need to a general API for such things.

Reply:
Another case to consider is styling for the column header.

Reply:
As we discussed in the team meeting, there will be an API to provide static styles for a column. Styling for the header can be provided with normal light DOM styling on the table since it's slotted content. It would be good to mention both of these in the spec.

Reply:
One other corner case to consider documenting: should we describe how a cell's height is computed? Are there any restrictions on what contents can go in a cell (e.g. all cells should have the same height) or things to know about what layout the table will apply to its cells?

Reply:
I'm not sure we need to get into the details for how a cell's height is computed. Ultimately, this is figured out when considering the heights across all cells in a row (maybe...we could decide that row height is made explicit via some other API, and the cellTemplate of a column has no say in things).